cpp/src/arrow/util/macros.h

```diff
@@ -36,40 +36,63 @@
   TypeName& operator=(TypeName&&) = default
 #endif
 
-#define ARROW_UNUSED(x) (void)(x)
-#define ARROW_ARG_UNUSED(x)
+// With ARROW_PREDICT_FALSE, GCC and clang can be told that a certain branch is
+// not likely to be taken (for instance, a CHECK failure), and use that information in
+// static analysis. Giving the compiler this information can affect the generated code
+// layout in the absence of better information (i.e. -fprofile-arcs). [1] explains how
+// this feature can be used to improve code generation. It was written as a positive
+// comment to a negative article about the use of these annotations.
 //
-// GCC can be told that a certain branch is not likely to be taken (for
-// instance, a CHECK failure), and use that information in static analysis.
-// Giving it this information can help it optimize for the common case in
-// the absence of better information (ie. -fprofile-arcs).
+// ARROW_COMPILER_ASSUME allows the compiler to assume that a given expression
+// is true, without evaluating it, and to optimise based on this assumption [2].
+// If this condition is violated at runtime, the behavior is undefined. This can
+// be useful to generate both faster and smaller code in compute kernels. Different
+// optimisers are likely to react differently to this annotation. It should be used
+// with care [3] when we can prove by some means that the assumption is guaranteed
+// to always hold.
```

Comment:
Also @mapleFU 's link about possible pessimizations due to misuse of assumptions?

Reply:
Yeah but let benchmark decides, I think it would be useful

Reply:
I greatly expanded the comments here. Pushing now.